src/Functions/LowerUpperUTF8Impl.h

```diff
@@ -102,179 +42,37 @@ struct LowerUpperUTF8Impl
-    }
-
-    static void vectorFixed(const ColumnString::Chars &, size_t, ColumnString::Chars &)
-    {
-        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Functions lowerUTF8 and upperUTF8 cannot work with FixedString argument");
-    }
-
-    /** Converts a single code point starting at `src` to desired case, storing result starting at `dst`.
-     *    `src` and `dst` are incremented by corresponding sequence lengths. */
-    static bool toCase(const UInt8 *& src, const UInt8 * src_end, UInt8 *& dst, bool partial)
-    {
-        if (src[0] <= ascii_upper_bound)
+        res_data.resize(data.size());
+        res_offsets.resize_exact(offsets.size());
+        size_t curr_offset = 0;
+        for (size_t i = 0; i < offsets.size(); ++i)
         {
-            if (*src >= not_case_lower_bound && *src <= not_case_upper_bound)
-                *dst++ = *src++ ^ flip_case_mask;
+            const auto * pos =  reinterpret_cast<const char *>(&data[offsets[i-1]]);
+            size_t size = offsets[i] - offsets[i-1];
+
+            icu::UnicodeString input(pos, static_cast<int32_t>(size), "UTF-8");
+            if constexpr (upper)
+                input.toUpper();
             else
-                *dst++ = *src++;
-        }
-        else if (src + 1 < src_end
-            && ((src[0] == 0xD0u && (src[1] >= 0x80u && src[1] <= 0xBFu)) || (src[0] == 0xD1u && (src[1] >= 0x80u && src[1] <= 0x9Fu))))
-        {
-            cyrillic_to_case(src, dst);
-        }
-        else if (src + 1 < src_end && src[0] == 0xC2u)
-        {
-            /// Punctuation U+0080 - U+00BF, UTF-8: C2 80 - C2 BF
-            *dst++ = *src++;
-            *dst++ = *src++;
-        }
-        else if (src + 2 < src_end && src[0] == 0xE2u)
-        {
-            /// Characters U+2000 - U+2FFF, UTF-8: E2 80 80 - E2 BF BF
-            *dst++ = *src++;
-            *dst++ = *src++;
-            *dst++ = *src++;
-        }
-        else
-        {
-            size_t src_sequence_length = UTF8::seqLength(*src);
-            /// In case partial buffer was passed (due to SSE optimization)
-            /// we cannot convert it with current src_end, but we may have more
-            /// bytes to convert and eventually got correct symbol.
-            if (partial && src_sequence_length > static_cast<size_t>(src_end - src))
-                return false;
+                input.toLower();
 
-            auto src_code_point = UTF8::convertUTF8ToCodePoint(src, src_end - src);
-            if (src_code_point)
-            {
-                int dst_code_point = to_case(*src_code_point);
-                if (dst_code_point > 0)
-                {
-                    size_t dst_sequence_length = UTF8::convertCodePointToUTF8(dst_code_point, dst, src_end - src);
-                    assert(dst_sequence_length <= 4);
+            String output;
+            input.toUTF8String(output);
+            const char * end = find_first_symbols<'\0'>(output.data(), output.data() + output.size());
```

Comment:
I found sometimes the output has extra '\0's by  the end. So we need to get content before the first '\0'. We don't need to worry about the issue of remining output because it won't happen for a valid UTF8 string. It we met an invalid UTF8 string, the behavior is not defined and current implementation is also reasonable.